PerimeterXModule/PxModule.cs

```diff
@@ -1,629 +1,671 @@
+					{
+						pxContext.LoginCredentialsFields = loginCredentialsFields;
+                    }
+                }
+
+				// validate using risk cookie
+				IPxCookie pxCookie = PxCookieUtils.BuildCookie(config, pxContext.PxCookies, cookieDecoder);
+				pxContext.OriginalToken = PxCookieUtils.BuildCookie(config, pxContext.OriginalTokens, cookieDecoder);
+				if (!PxCookieValidator.Verify(pxContext, pxCookie))
+				{
+					// validate using server risk api
+					PxS2SValidator.VerifyS2S(pxContext);
+				}
+
+				HandleVerification(application);
+			}
+			catch (Exception ex) // Fail-open approach
+			{
+				PxLoggingUtils.LogError(string.Format("Module failed to process request in fault: {0}, passing request", ex.Message));
+				pxContext.PassReason = PassReasonEnum.ERROR;
+				PostPageRequestedActivity(pxContext);
+			}
+		}
+
+		private static string ByteArrayToHexString(byte[] input)
+		{
+			StringBuilder sb = new StringBuilder(input.Length * 2);
+			foreach (byte b in input)
+			{
+				sb.Append(PxConstants.HEX_ALPHABET[b >> 4]);
+				sb.Append(PxConstants.HEX_ALPHABET[b & 0xF]);
+			}
+			return sb.ToString();
+		}
+
+		private void HandleVerification(HttpApplication application)
+		{
+			PxModuleConfigurationSection config = (PxModuleConfigurationSection)ConfigurationManager.GetSection(PxConstants.CONFIG_SECTION);
+			bool verified = blockingScore > pxContext.Score;
+
+			PxLoggingUtils.LogDebug(string.Format("Request score: {0}, blocking score: {1}, monitor mode status: {2}.", pxContext.Score, blockingScore, config.MonitorMode == true ? "On" : "Off"));
+
+			if (verified)
+			{
+				if (config.MonitorMode)
+				{
+					PxLoggingUtils.LogDebug("Monitor Mode is activated. passing request");
+				}
+
+				HandleCredentialsIntelligence(application, config);
+
+				PxLoggingUtils.LogDebug(string.Format("Valid request to {0}", application.Context.Request.RawUrl));
+				PostPageRequestedActivity(pxContext);
+			}
+			else
+			{
+				PxLoggingUtils.LogDebug(string.Format("Invalid request to {0}", application.Context.Request.RawUrl));
+				PostBlockActivity(pxContext, config);
+			}
+
+			SetPxhdAndVid(pxContext);
+			// If implemented, run the customVerificationHandler.
+			if (customVerificationHandlerInstance != null)
+			{
+				customVerificationHandlerInstance.Handle(application, pxContext, config);
+			}
+			// No custom verification handler -> continue regular flow
+			else if (!verified && !pxContext.MonitorRequest)
+			{
+				BlockRequest(pxContext, config);
+				application.CompleteRequest();
+			}
+		}
+
+		private static void SetPxhdAndVid(PxContext pxContext)
+		{
+
+			if (!string.IsNullOrEmpty(pxContext.Pxhd))
+			{
+				string pxhd = PxConstants.COOKIE_PXHD_PREFIX + "=" + pxContext.Pxhd + "; path=/";
+				pxContext.ApplicationContext.Response.AddHeader("Set-Cookie", pxhd);
+			}
+		}
+
+		private void HandleCredentialsIntelligence(HttpApplication application, PxModuleConfigurationSection config)

```

Comment:
This function name isn't clear -- is it the credentials extraction portion? The additional_s2s activity portion? The general name suggests that it does all of it, so I'd change the name to something more specific -- `AddCredentialIntelligenceHeadersToRequest`?